### ISSUE TITLE
RF idea: sparse-file cache information to be stored per-cached file (instead of a singular `/cache`)

### ISSUE THREAD

**Issue by yarikoptic** [intake/filesystem_spec#556]
Related to #555 (as if the layout to change) and may be other issues (e.g. #553 ) on closing/reopening cache.  I could be totally wrong, but ATM I think fsspec stores cache information (e.g. which blocks of the sparse file were already fetched etc?) in a singular `/cache` file within the cache directory.  I wonder, if e.g. cache is to store thousands of files - how efficient dumping of the cache information would be? 
Ideally cache meta data should be updated "on disk" as soon as a new block for a file is fetched/stored into the cache sparse file, so there would be no need for any additional metadata dump upon `close` or `__del__`, but likely it would be an expensive operation if it is the entire dump of metadata for all files to be dumped at once.

So I wonder if it is worth adopting the same approach as `rclone` took for its `mount` operation.  It has two directories in the cache -- `vfs` and `vfsMeta` where in `vfs` are the actual sparse files, and in `vfsMeta` (I guess) under the same key path, all the information about that particular cached file, e.g.

```shell
$> ls -ld /home/yoh/.cache/rclone/vfs*/s3_us_east_2/dandiarchive/largeuploads/blobs/f5d/aac/79d9c6c1ba2439200d45680f1cb941087750a000e484ea2973232709d2 
-rw------- 1 yoh yoh 3628888939868 Feb 18 20:57 /home/yoh/.cache/rclone/vfs/s3_us_east_2/dandiarchive/largeuploads/blobs/f5d/aac/79d9c6c1ba2439200d45680f1cb941087750a000e484ea2973232709d2
-rw------- 1 yoh yoh           261 Mar  3 11:14 /home/yoh/.cache/rclone/vfsMeta/s3_us_east_2/dandiarchive/largeuploads/blobs/f5d/aac/79d9c6c1ba2439200d45680f1cb941087750a000e484ea2973232709d2
```

such layout also makes it easy with regular `rm` to prune cache for no longer desired files etc.

**Comment by martindurant** [comment]
I don't believe people are hitting this limit yet, but your model sounds interesting. 
Note that "simplecache", by far the most common caching mechanism, doesn't store metadata and so is immune to this problem.

**Comment by martindurant** [comment]
Do you really have a single file that big??

**Comment by yarikoptic** [comment]
yeap -- you can just append s3:// there before `/dandiarchive` in the path (do not want to paste in full to avoid some crawly causing unneeded data transfer). 
background: Even if not that many files that big yet -- we have amassed over 200TB of data (much of it on S3) indexed by git-annex (http://datasets.datalad.org/). And more of similarly large files are to come (we do have a good number of 100s of GBs large files in that dandiarchive bucket). My "greedy" use case is to provide FUSE filesystem on top of datalad datasets (git annex repositories, [ugly prototype is here](https://github.com/datalad/datalad-fuse/blob/master/datalad_fuse/fuse_.py)) so we could access only needed portions of the files (e.g. for metadata extraction etc).  That is why my interest in fsspec's "blockcache" to avoid some ad-hoc implementation of what fsspec already does and desire to "scale" in number of cached files (#555)

**Comment by martindurant** [comment]
Have you already tried fsspec's own FUSE implementation?

For indexing or otherwise scanning massive files for partial access, you might be interested in https://github.com/intake/fsspec-reference-maker . The only example we have so far is explicitly for massive HDF5 files - assuming they are netCDF-like (because the target is xarray, but actually zarr doesn't require this).

**Comment by martindurant** [comment]
Hmm, datalad seems to have a fair bit of overlap with [Intake](https://intake.readthedocs.io/en/latest/), or could be one of the datasources which Intake presents as a catalogue. Intake does not store any data of its own.

**Comment by yarikoptic** [comment]
> Have you already tried fsspec's own FUSE implementation?

I have "consulted" it ;) Even started to hack something to make it usable as `python -m fsspec.fuse --fs http ... `... ;)

My then realized that my use case is quite different - I would like to simply pass-through access to some (already present locally) files, and only handle the other (pretty much "broken symlinks of git-annex" ATM but later also its "unlocked files"), also I am mostly interested in read-only access ATM, and later might interface multiple file systems (e.g. http + s3) depending on where any given git-annex'ed files can be obtained from.

> ... datalad seems to have a fair bit of overlap with Intake ...

Yes, indeed -- for the aspects of "data distribution" or "virtualized filesystem" or ... But lots of differences though as well to warrant the existence of both ;-) 
Indeed metadata from any DataLad dataset could be exported so it could be then provided as a manifest for data availability etc to be used by intake.  We are working ATM on better/tighter integration with metadata management, but even our existing implementation provides lots of metadata for search/download/etc.

**Comment by martindurant** [comment]
At the very least, I think there ought to be an Intake catalogue driver fr datalad, so that entries within it (or the whole thing) could be included among other Intake datasets.

**Comment by yarikoptic** [comment]
A thought came up that likely storing cache information per file could at large provide remedy also for #1107 AKA #1126 -- concurrency issue, since then locking could be implemented at per target file level thus avoiding effective "serialization" of operations if working on a single monolythic `cache`.

**Comment by martindurant** [comment]
Yes, you are probably right. I think the whole cache stuff could do with some care and design...
